slick/src/main/scala/slick/util/ManagedArrayBlockingQueue.scala

```diff
@@ -1,87 +1,101 @@
 package slick.util
 
-import java.util.concurrent.{ArrayBlockingQueue, TimeUnit, BlockingQueue}
+import java.util.concurrent.{BlockingQueue, TimeUnit}
 import java.util.concurrent.locks._
-import java.util._
+import java.util
+import java.util.function.Supplier
+
+import slick.util.AsyncExecutor.{HighPriority, LowPriority, MediumPriority, Priority}
 
 /** A simplified copy of `java.util.concurrent.ArrayBlockingQueue` with additional logic for
   * temporarily rejecting elements based on the current size. All features of the original
   * ArrayBlockingQueue have been ported, except the mutation methods of the iterator. See
   * `java.util.concurrent.ArrayBlockingQueue` for documentation. */
-abstract class ManagedArrayBlockingQueue[E >: Null <: AnyRef](capacity: Int, fair: Boolean = false) extends AbstractQueue[E] with BlockingQueue[E] { self =>
+abstract class ManagedArrayBlockingQueue[E >: Null <: AnyRef](maximumInUse: Int, capacity: Int, fair: Boolean = false)
+  extends util.AbstractQueue[E]
+  with BlockingQueue[E]
+  with Logging { self =>
 
-  /** Determine if the item should be accepted at the current time. */
-  protected[this] def accept(item: E, size: Int): Boolean
+  protected[this] def priority(item: E): Priority
 
-  private[this] val items = new Array[AnyRef](capacity)
   private[this] val lock = new ReentrantLock(fair)
   private[this] val notEmpty = lock.newCondition
   private[this] val notFull = lock.newCondition
-  private[this] var takeIndex, putIndex, count = 0
 
   private[this] def checkNotNull(v: AnyRef): Unit = if (v == null) throw new NullPointerException
 
-  private[this] def inc(i: Int): Int = if(i+1 == items.length) 0 else i+1
+  private[this] val itemQueue = new InternalArrayQueue[E](2*capacity)
+  private[this] val highPrioItemQueue = new InternalArrayQueue[E](capacity)
+
+  private[this] def counts = (if (paused) 0 else  itemQueue.count) + highPrioItemQueue.count
 
-  private[this] def dec(i: Int): Int = (if(i == 0) items.length else i) - 1
+  /**
+    * The number of low/medium priority items in use
+    */
+  private[this] var inUseCount = 0
 
-  private[this] def itemAt(i: Int): E = items(i).asInstanceOf[E]
+  /**
+    * Because the connection in use count can be incremented from two call sites (but always by the threadpool's worker thread)
+    * we protect it by an ThreadLocal, so it will only get incremented once
+    */
+  private[this] val inUseCountThreadLocal = ThreadLocal.withInitial(new Supplier[Boolean] {
```

Comment:
Would it make sense to manage this state in `PrioritizedRunnable` instead so we don't need a `ThreadLocal`?